log/logger.go

```diff
@@ -42,34 +121,206 @@ func File(path string) (logger *Logger, loggerClose func(), err error) {
-	loggerClose = func() {
-		if err := file.Close(); err != nil {
-			panic(err)
+
+	writer = io.MultiWriter(parent, file)
+
+	return writer, nil
+}
+
+// spawningHandler is a structural logging handler which spawns a new log file if one of the given log file spawners triggers.
+type spawningHandler struct {
+	// writer holds the writer to write the output.
+	writer io.Writer
+
+	// attributes holds the attributes handed to the logger.
+	attributes map[AttributeKey]string
+
+	// logFileSpawners holds the spawners responsible for spawning a new log file.
+	logFileSpawners []logFileSpawner
+
+	// flags holds the flags deciding which fields are logged.
+	flags Flags
+}
+
+// newSpawningHandler returns a new spawning handler.
+func newSpawningHandler(writer io.Writer, flags Flags) *spawningHandler {
+	return &spawningHandler{
+		writer: writer,
+
+		attributes: map[AttributeKey]string{},
+
+		flags: flags,
+	}
+}
+
+// Clone returns a copy of the object.
+func (h *spawningHandler) Clone() (clone *spawningHandler) {
+	return &spawningHandler{
+		writer: h.writer,
+
+		attributes: maps.Clone(h.attributes),
+
+		logFileSpawners: slices.Clone(h.logFileSpawners),
+	}
+}
+
+var _ slog.Handler = (*spawningHandler)(nil)
+
+// Enabled reports whether the handler handles records at the given level.
+func (h *spawningHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return true
+}
+
+// Handle handles the Record.
+func (h *spawningHandler) Handle(ctx context.Context, record slog.Record) error {
+	if h.flags&FlagDate != 0 {
+		fmt.Fprint(h.writer, record.Time.Format("2006/01/02"))
+		fmt.Fprint(h.writer, " ")
+	}
+	if h.flags&FlagTime != 0 {
+		fmt.Fprint(h.writer, record.Time.Format("15:04:05"))
+		fmt.Fprint(h.writer, " ")
+	}
+
+	fmt.Fprintln(h.writer, record.Message)
+
+	return nil
+}
+
+// WithAttrs returns a new Handler whose attributes consist of both the receiver's attributes and the arguments.
+// The Handler owns the slice: it may retain, modify or discard it.
+func (h *spawningHandler) WithAttrs(attributes []slog.Attr) slog.Handler {
+	// Collect attributes.
+	for _, attribute := range attributes {
+		h.attributes[AttributeKey(attribute.Key)] = attribute.Value.String()
+	}
+
+	newHandler := h.Clone()
+
+	// Check if we need to spawn a new log file.
+	for i, spawner := range h.logFileSpawners {
+		if spawner.NeedsSpawn(h.attributes) {
+			logFilePath := spawner.FilePath(h.attributes)
+			writer, err := newLogWriter(h.writer, logFilePath)
+			if err != nil {
+				fmt.Fprintf(h.writer, "ERROR: cannot create new handler: %s\n", err.Error())
+			}
+
+			logMessage := fmt.Sprintf("Spawning new log file at %s", logFilePath)
+			h.Handle(context.Background(), slog.NewRecord(time.Now(), slog.LevelInfo, logMessage, 0))
+
+			newHandler.writer = writer
+			newHandler.logFileSpawners = slices.Delete(newHandler.logFileSpawners, i, i+1) // The currently triggered log file spawner must not be part of the new handler as it would trigger again and again.
```

Comment:
can we delete  the entry from the slice while we are iterating over it?

Reply:
we iterate over `h.logFileSpawners` and delete from the clone `newHandler.logFileSpawners`